### DISCUSSION TITLE
using Subject metadata in torchio.transforms

### DISCUSSION THREAD

**Discussion by Jarartur** [TorchIO-project/torchio#698] [Q&A] [answered]
Hello,
I am using a 2D histology dataset from [here](https://anhir.grand-challenge.org/) and my task is registration.
I am loading my image pairs and saving them as two `ScalarImages` in a `Subject`. All images have the same pixel size in mm so to do registration I need to pad them to be the same shape.
I searched through the docs but I am still confused as to using other metadata from a `Subject` instance. Can I for example use it like that:
```python
data = pd.read_csv('summary_file.csv')
subject = tio.Subject(
        src=tio.ScalarImage(data['src_path'][0]),
        trg=tio.ScalarImage(data['trg_path'][0]),
        desired_shape=data['shapes'][0]
        )
transform = tio.Compose([
    tio.CropOrPad(subject['desired_shape'])
])
augmented_subject = tranform(subject)
```

This example works but only because I am using one subject and not a list. My question is can I somehow reference `desired_shape` key in every iteration on a list of subjects.

**Comment by fepegar** [comment, marked as answer]
Do you need a different `desired_shape` for each subject? If you do, maybe you can create a custom transform.

```python
def CustomCrop(subject):
    crop_pad = tio.CropOrPad(subject['desired_shape'])
    return crop_pad(subject)
```

**Reply by Jarartur** [reply]
By looking over the source code I thought that into transform goes already subject['src'] and subject['trg'], but If the whole subject can go in, then I guess that would solve my problem.
I will try this.

**Reply by Jarartur** [reply]
It works like a charm. Really thank you for the answer!
Now I see I misinterpreted something in `tio.Compose`, because in fact it applies the function on the whole subject and the transform function decides on which subject parameters to work on.

**Reply by fepegar** [reply]
No worries!

**Reply by Jarartur** [reply]
I tested this further but unfortunately `tio.CropOrPad` doesn't accept different shapes of images in one subject giving me:
```
RuntimeError: More than one value for "spatial_shape" found in subject images:
{'src': (387, 512, 1), 'trg': (389, 512, 1)}
```
I can bypass this by using a custom function for padding completely though so I think I will do that as there is stil much benefit in using torchio's queue in patch-based pipieline.

**Reply by fepegar** [reply]
I suppose your images occupy the same physical space. You can use e.g. `Resample('src')` so they have the same shape before crop/pad. Have you gone through the [tutorials](https://torchio.readthedocs.io/quickstart.html#tutorials)?

**Reply by Jarartur** [reply]
Yes, but my use case is a bit of an edge case because I have a whole piepieline to do manual affine registration before learning.
All images have the same physical pixel size, but different shapes and my current set of affine matrices are calculated on padded images to the same shape (as to keep the original pixel size).
Maybe I incorrectly understood what `Resample` really is but I don't think it can be used to make images the same spatial shape before padding because it would change the physical dimensions of pixels in the pair of images. _(Unfortunately organizers provide only .jpg/.png images so I operate on limited metadata)_

**Reply by fepegar** [reply]
> I don't think it can be used to make images the same spatial shape before padding because it would change the physical dimensions of pixels in the pair of images.

That is what `Resize` does, which I try not to recommend. `Resample` does not change the resolution. Some more information on resampling can be found on the [SimpleITK docs](https://simpleitk.readthedocs.io/en/master/fundamentalConcepts.html#resampling).

**Reply by Jarartur** [reply]
Ah, yes. You are right and I misunderstood what `Resample` really is. It in fact padded my images in some cases.
Unfortunately to calculate affine matrices I padded them manually in a different way. When the spatial dimensions didn't match I divided the padding by two and applied it to two sides of an image. `Resample` as I see now pads the smaller image on one side to match the dimensions which is not compatible with my current preprocessing.
Also in cases when one image was bigger in one dimension and other image in the other dimension I padded them both to reduce interpolation on them.

**Reply by fepegar** [reply]
Conceptually, it is a bit different. Images are not padded or cropped: their values a re-sampled in the output/target/reference space, and that's why both images end up having the same shape. "Padded" and "cropped" pixels are just the ones without correspondence between the images within their bounds. If you share a pair of images, maybe we can advise further.

**Reply by Jarartur** [reply]
That being said I learned a lot on how torchio works and I think I can still employ it with my custom preprocessing. Also, in the future I will certainly look more into how to preprocess them using more standard ways.

**Reply by Jarartur** [reply]
![image](https://user-images.githubusercontent.com/45856425/138698134-35ef4843-f201-48d8-9b15-388a819f8470.png)

Here is the example I meant after applying `Resample('src')`
And from what I understand in case where 'src' shape would be i.e. 14000x7000 and 'trg' shape 8000x 15000 `Resample('src')` would resample 'trg' image in 14000x7000 shape which would greatly reduce its information on tissue structure.

**Reply by fepegar** [reply]
`CropOrPad` is designed to be safe against misaligned medical images. In your case, images are misaligned anyway because that's the point of the task, so this issue can be ignored, I suppose. It is a little bit difficult to discuss these issues because you are not using medical images (which have certain associated spatial metadata).

My first question is: why do the width/height ratios of both images, 14/7 and 8/15, differ so much?

**Reply by Jarartur** [reply]
It happens when one image is rotated aprox. 90 degrees which is in the range of the task. Authors then crop the image to conserve space as one of those images currently takes 75-100 Mb.
As the model would require enormous receptive field for a 90-degree rotation I pre-align them manually and then make non-rigid registration on the patches of data.
Also, I can't simply save them aligned as there are ~200 images but 480 pairs so I need to pre-algin them on load time.
So, my problems come from the non-standard dataset attributes, and they are solved by custom loading them.

**Reply by Jarartur** [reply]
But I think most of my problems will be solved by creating a custom transform which shouldn't be that hard to do as I need to adapt my previous code to make use of the `subject` object. If I can use things like OpenCV in the `tio.Compose` then it should be straight forward now that I know how everything works.
